plugins/source/metrics.go

```diff
@@ -123,3 +130,51 @@ func (s *Metrics) TotalResourcesAtomic() uint64 {
 	}
 	return total
 }
+
+func (s *Metrics) MarkStart(table *schema.Table, clientID string) {
+	s.TableClient[table.Name][clientID].Mutex.Lock()
+	defer s.TableClient[table.Name][clientID].Mutex.Unlock()
+
+	s.TableClient[table.Name][clientID].StartTime = time.Now()
+}
+
+// If the table is a top-level table, we need to mark all of its descendents as 'done' as well.
+// This is because, when a top-level table is empty (no resources), its descendants are never actually
+// synced.
+func (s *Metrics) MarkEnd(table *schema.Table, clientID string) {
+	if table.Parent == nil {
+		s.markEndRecursive(table, clientID)
+	} else {
+		s.TableClient[table.Name][clientID].Mutex.Lock()
+		defer s.TableClient[table.Name][clientID].Mutex.Unlock()
+		s.TableClient[table.Name][clientID].EndTime = time.Now()
+	}
+}
+
+func (s *Metrics) markEndRecursive(table *schema.Table, clientID string) {
+	// We don't use defer with Unlock(), because we want to unlock the mutex as soon as possible.
+	s.TableClient[table.Name][clientID].Mutex.Lock()
```

Comment:
Maybe we should lift the locking here to `MarkEnd`? I'm not sure we'd like to have interleaving locks and updating the times is not an expensive operation.